packages/flutter/lib/src/widgets/splash.dart

```diff
@@ -0,0 +1,621 @@
+          return null;
+        }
+        toPath.add(toParent);
+        to = toParent;
+      }
+    }
+    assert(identical(from, to));
+
+    final Matrix4 transform = Matrix4.identity();
+    final Matrix4 inverseTransform = Matrix4.identity();
+
+    for (int index = toPath.length - 1; index > 0; index -= 1) {
+      toPath[index].applyPaintTransform(toPath[index - 1], transform);
+    }
+    for (int index = fromPath.length - 1; index > 0; index -= 1) {
+      fromPath[index].applyPaintTransform(fromPath[index - 1], inverseTransform);
+    }
+
+    final double det = inverseTransform.invert();
+    return det != 0 ? (inverseTransform..multiply(transform)) : null;
+  }
+
+  /// Called when the user input that triggered this feature's appearance was confirmed.
+  ///
+  /// Typically causes the [Splash] to propagate faster across the surface.
+  /// By default this method does nothing.
+  void confirm() {}
+
+  /// Called when the user input that triggered this feature's appearance was canceled.
+  ///
+  /// Typically causes the [Splash] to gradually disappear.
+  /// By default this method does nothing.
+  void cancel() {}
+
+  /// {@macro flutter.widgets.splash.SplashController}
+  final SplashController controller;
+
+  /// The render box whose visual position defines the splash effect's
+  /// frame of reference.
+  final RenderBox referenceBox;
+
+  /// Called when the splash is no longer visible on the material.
+  final VoidCallback? onRemoved;
+
+  /// If asserts are enabled, this value tracks whether the feature has been disposed.
+  ///
+  /// Ensures that [dispose] is only called once, and [paint] is not called afterward.
+  bool debugDisposed = false;
+
+  /// A (typically translucent) color used for this [Splash].
+  Color get color => _color;
+  Color _color;
+  set color(Color value) {
+    if (value == _color) {
+      return;
+    }
+    _color = value;
+    controller.markNeedsPaint();
+  }
+
+  /// A [ShapeBorder] that may optionally be applied to the [Splash].
+  ShapeBorder? get customBorder => _customBorder;
+  ShapeBorder? _customBorder;
+  set customBorder(ShapeBorder? value) {
+    if (value == _customBorder) {
+      return;
+    }
+    _customBorder = value;
+    controller.markNeedsPaint();
+  }
+
+  /// Draws a [Splash] on the provided [Canvas].
+  ///
+  /// The [transform] argument is the [Matrix4] transform that typically
+  /// shifts the coordinate space of the canvas to the space in which
+  /// the circle is to be painted.
+  ///
+  /// If a [customBorder] is provided, then it (along with the [textDirection])
+  /// will be used to create a clipping path.
+  ///
+  /// Otherwise, the [clipCallback] clips the splash to a [RRect] (created by
+  /// applying the [borderRadius] to its result).
+  ///
+  /// If both [customBorder] and [clipCallback] are null, no clipping takes place.
+  ///
+  /// For examples on how the function is used, see [InkSplash] and [InkRipple].
+  @protected
+  void paintCircle({
```

Comment:
This method feels like a utility method that should be moved to a subclass.

Reply:
Yes, I believe we share the same understanding here.

As of now, the details regarding how to move this method to a subclass in a way that's intuitive—and hopefully non-breaking—aren't super clear. I may need some time to think this one over.

Reply:
Right now I feel stuck between 2 possible solutions:
- Keep it as `Splash.paintCircle()` (renamed with no other changes)
- Separate it into a `mixin` (an ideal solution, but isn't supported by a [data-driven fix](https://github.com/flutter/flutter/blob/master/docs/contributing/Data-driven-Fixes.md#replacedby))

The most recent commit reverts the changes to that method. Perhaps the best course of action is to not touch `paintInkCircle` for now, and then file a separate PR entirely focused on deprecating & replacing the method once this one lands.

Reply:
@Piinks self-ping to review this thread

Reply:
Followed up on this in the latest review. 👍